### ISSUE TITLE
Replicating DeepMind results

### ISSUE THREAD

**Issue by Kaixhin** [Kaixhin/Rainbow#15]
As of 5c252ea, this repo has been checked over several times for discrepancies, but is still unable to replicate DeepMind's results. This issue is to discuss any further points that may need fixing.

- [x] Should the [loss](https://github.com/Kaixhin/Rainbow/blob/master/agent.py#L82) be averaged or summed over the minibatch?
- [x] Should noisy network updating use independent noise per transition in the batch [[v1](https://arxiv.org/abs/1706.10295v1)] or the same noise but another noise sample for action selection [[v2](https://arxiv.org/abs/1706.10295v2)]?
- [x] Is the max priority over all time, or just from the current buffer (may be the former)? Results and paper indicate former.
- [x] Are priorities added as δ, or δ + ε (ε may not be needed with a KL loss)? One single ablation run indicates adding ε causes performance to drop more at end of training. δ + ε shouldn't be needed with a KL loss,
- [x] Most people implement PER by adding priorities already multiplied by α, but the maths indicates that the raw values should be stored and sampling should be done with respect to everything to the power of α? α isn't changed here - so not an issue.

Space Invaders (averaged losses):
![newplot](https://user-images.githubusercontent.com/991891/37013022-a7c76d44-20c5-11e8-9c74-bcf82a6de843.png)

Space Invaders (summed losses):
![newplot](https://user-images.githubusercontent.com/991891/36953006-c8ba2758-1fe4-11e8-9b08-249992d4d5ad.png)

**Comment by stringie** [comment]
I believe I read somewhere that the loss should be the max over the minibatch, but I think that sum should be just as fine.

**Comment by Kaixhin** [comment]
Asked Matteo a few questions, placed his responses [here](https://github.com/Kaixhin/Rainbow/wiki/Matteo's-Notes), and will do some new runs based on this new information.

**Comment by Kaixhin** [comment]
Trying a new game - Frostbite - which already seems to work fine! Space Invaders results are still along previous results, but it takes longer to see if it makes progress to the same scores. 

![newplot](https://user-images.githubusercontent.com/991891/37379526-34d93858-270a-11e8-9bd0-29ad68aecdf7.png)

**Comment by stringie** [comment]
Have you tried to see how it performs on breakout? I'm really curious.

**Comment by Kaixhin** [comment]
I have limited resources so not yet but plan to eventually. If that doesn't work I'll add the fire-on-reset wrapper to the environment and try again.

**Comment by stringie** [comment]
I understand. It really is power hungry to run.

**Comment by stringie** [comment]
Have you seen the paper on Distributed Prioritized Experience Replay? The results look amazing relative to what Rainbow can achieve. I wonder if it can also somehow be integrated in your project. Here's the paper: https://arxiv.org/pdf/1803.00933.pdf

**Comment by Kaixhin** [comment]
@stringie this issue is to track "replicating DeepMind results", so this is not relevant here. Anyway, adding extra components is a) currently not within the scope of this project b) unhelpful when results from the original Rainbow paper still cannot be replicated.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin is it possible to try individual components of Rainbow and check whether the performances are similar to the respective papers?

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari if I had written the code in a modular fashion that allowed these kinds of experiments. There's plenty of things I've not needed to include by just aiming for the full Rainbow model, and I don't have the capacity to refactor the code in this manner at the moment.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin  So I recently incorporated the PER from your code to a vanilla DQN and ran it for Breakout. It seems, PER is where the issue lies. As it is not behaving as per the claims in Prioritized Experience Replay paper (slightly worse than DQN while training, although more stable).

The experiments were run for Replay memory size of Replay_Mem_size_in_DQN_paper/5 i.e. 200k. 
But not sure where the issue lies. 

Also, I did not quite understand the function of attribute 'n' of a ReplayMemory object in your code in memory.py and henceforth the logic behind line 109 and 116 in memory.py.

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari that's useful to know, thanks. Is any code public so that I can have a look at it?

`n` comes from [n-step backups](http://incompleteideas.net/book/ebook/node73.html), so if you set `n = 1` you should recover most of the original algorithms.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin  Thanks for the clarification regarding `n`.  Give me a few days, I shall make the code public.

Till then, the basic code I used was from https://github.com/hengyuan-hu/rainbow.git.

The way they store the samples is slightly different, for ex. a sample will contain both the current as well as next state. But nevertheless, that should not distort the logic. 

Also, not sure how scaling down the Replaymemory would affect the relative performances for modules related to experience replay (any insights over this?). I think the performance difference might become lesser between vanilla DQN and PER DQN, but vanilla DQN >  PER DQN should not happen (?).

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari actually PER can decrease performance on a few games. If you check Figure 6 in the appendix of the Rainbow paper you will see that removing PER from Rainbow actually improves performance in Breakout. In this case you would want to test against a game that the paper shows is clearly positively effected, such as Yars' Revenge. For normal ER I would say that /5 may be OK, but with PER I'm less certain.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin I think while we are referring to vanilla DQN and addition of PER to it (since we are trying to check PER as a module), we might want to focus on  [PER paper](https://arxiv.org/abs/1511.05952). Which mentions improvement in Table 6 of its appendix. Or no? 

If this is the case, then scaling down should affect or not (again in relative terms only)?

Because, I believe, when we are removing PER from Double, Duelling, etc. DQN, it might become too complex to draw a conclusion from.

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari yes sorry the repo you pointed to does have most of the parts of Rainbow but if you are trying to do a comparison against just the vanilla DQN then it's best to check Figure 7 of the original PER paper, as the learning curves (which have been smoothed) are the most informative. The **big** caveat is that the double DQN paper introduced a new set of hyperparameters for the DQN algorithm which work better for the double DQN, but I'm pretty sure the baseline results shown in Figure 7 don't use these improved hyperparameters (whereas the PER model does). Nevertheless I think the PER learning curves should be unique so Frostbite or Space Invaders are good games to test if PER is working correctly.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin I think I can check for Space Invaders (in a few days) after incorporating `n` and keeping `n = 1`. However, on a second run for Breakout, the best average (over 10 episodes) val reward for PER : 387 and for vanilla DQN : 384. The improvement is definitely not significant.
Note :1) ReplayMem size of 200k (1e6/5). 
          2) Also, not running for more than 25M (this should not be an issue).
          3) Sharing the Training curve seems illogical as of now.


![image](https://user-images.githubusercontent.com/15244719/38945547-58016c58-4369-11e8-9b38-b1c06c429102.png)

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari according to Figure 7 and other stats from the paper there isn't much difference between Breakout scores for all of the different methods, so it's difficult to draw conclusions from (the best are games where there should be a clear difference in learning). Especially as the published results can come from taking an average over several runs.

If you're still investigating then if you could also plot a) the max priority b) the priorities of the 32 samples per minibatch over time, then that might provide some clues if things are going wrong?

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin  Yep, the way they are taking averages is way different.  Taking average over a 100 episodes, that would definitely increase that value I am sure (law of large numbers, I think :P, as when we are taking average over 10 episodes only, even one or two rare poor performances are too many to bring that count lower).

Nevertheless, one should try it on different games as mentioned by you, right. 
So, basically we can't be sure as of now, whether there is a bug in PER or not, right?

**Comment by Kaixhin** [comment]
@Ashutosh-Adhikari averaging over more evaluation episodes would result in a better approximation to the true mean performance, it shouldn't bias the value either way. If anything it may result in smoother curves but the values should be comparable.

It's pretty hard to tell where the bug is (could even be in PyTorch if we're really unlucky), so yes it may not be in PER but it's one of the trickiest things to get right so it's most likely to be there. In my experience the best way to debug by running experiments is to pick options based on learning curves where there are clear differences between the options. Unit tests would also be a good idea, but depend on understanding the items correctly in the first place.

**Comment by Ashutosh-Adhikari** [comment]
@Kaixhin  Just added a pull request  to https://github.com/hengyuan-hu/rainbow.git for PER code inspired by your repo.

**Comment by Kaixhin** [comment]
Closing as results on Space Invaders are good and show a clear difference against previous methods:
![newplot](https://user-images.githubusercontent.com/991891/39676171-b2ad91fc-515d-11e8-88f7-440287b4bb96.png)

Enduro (still 2x reported, not sure why):
![newplot 2](https://user-images.githubusercontent.com/991891/39676185-01709424-515e-11e8-82d0-363e6b0ccfd0.png)

Frostbite:
![newplot 1](https://user-images.githubusercontent.com/991891/39676186-0d7db0ee-515e-11e8-8a81-855d2b9032d7.png)

**Comment by albertwujj** [comment]
Hi Kaixhin,

Are the hyperparams you used for these results currently the default ones in main.py?

Additionally, in the 'Human-level control' DQN paper, there is this note under Extended Table 4:

![image](https://user-images.githubusercontent.com/21957230/58908027-576d3e80-86c4-11e9-8978-bb08557cada3.png)

At 60 fps, 5 minutes is 18000 frames. This may be related to why your Enduro results are 2 times higher.

Thanks,
Albert

**Comment by Kaixhin** [comment]
@albertwujj if you check the [releases](https://github.com/Kaixhin/Rainbow/releases) you'll see that Enduro results match the paper now at just over 2k. The scores in general can be highly impacted by the total frame cap for the Atari games and the ε value used for ε-greedy during validation.
